File: 剑指offer/33.二叉搜索树的后序遍历序列.py
from typing import List


class Solution:
    def verifyPostorder(self, postorder: List[int]) -> bool:
        def help(i, j):
            if i >= j:
                return True

            p = i
            while postorder[p] < postorder[j]:
                p += 1
            m = p
            while postorder[p] > postorder[j]:
                p += 1
            return p == j and help(i, m - 1) and help(m, j - 1)

        return help(0, len(postorder) - 1)

File: 剑指offer/33.二叉搜索树的后序遍历序列.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool help(const vector<int> postorder, int left, int right)
    {
        if (left > right) {
            return true;
        }

        int i = left;
        while (postorder[i] < postorder[right]) {
            i++;
        }

        int mid = i;

        while (postorder[i] > postorder[right]) {
            i++;
        }

        return i == right && help(postorder, left, mid - 1) && help(postorder, mid, right - 1);
    }

    bool verifyPostorder(vector<int>& postorder)
    {
        return help(postorder, 0, postorder.size() - 1);
    }
};